datumaro/components/project.py

```diff
@@ -2,488 +2,2181 @@
+        if osp.isfile(src):
+            os.remove(src)
+            return
+
+        src += self.DIR_HASH_SUFFIX
+        if not osp.isfile(src):
+            raise UnknownRefError(obj_hash)
+
+        with open(src) as f:
+            src_meta = json.load(f)
+        for entry in src_meta:
+            entry_path = self.obj_path(entry['md5'])
+            if osp.isfile(entry_path):
+                os.remove(entry_path)
+
+        os.remove(src)
+
+class Tree:
+    # can be:
+    # - attached to the work dir
+    # - attached to a revision
+
+    def __init__(self, project: 'Project',
+            config: Optional[TreeConfig] = None,
+            rev: Optional[str] = None):
+        assert isinstance(project, Project)
+        assert not rev or project.is_ref(rev), rev
+
+        if not isinstance(config, TreeConfig):
+            config = TreeConfig(config)
+        if config.format_version != 2:
+            raise ValueError("Unexpected tree config version '%s', expected 2" %
+                config.format_version)
+        self._config = config
+
+        self._project = project
+        self._rev = rev
+
+        self._sources = ProjectSources(self)
+        self._targets = ProjectBuildTargets(self)
+
+    def save(self):
+        self.dump(self._config.config_path)
+
+    def dump(self, path):
+        os.makedirs(osp.dirname(path), exist_ok=True)
+        self._config.dump(path)
+
+    @property
+    def sources(self) -> ProjectSources:
         return self._sources
 
-    def _save_branch_project(self, extractor, save_dir=None):
-        if not isinstance(extractor, Dataset):
-            extractor = Dataset.from_extractors(
-                extractor) # apply lazy transforms to avoid repeating traversals
+    @property
+    def build_targets(self) -> ProjectBuildTargets:
+        return self._targets
+
+    @property
+    def config(self) -> Config:
+        return self._config
+
+    @property
+    def env(self) -> Environment:
+        return self._project.env
+
+    @property
+    def rev(self) -> str:
+        return self._rev
+
+    def make_dataset(self, target: Optional[str] = None) -> Dataset:
+        if not target:
+            target = 'project'
+
+        pipeline = self.build_targets.make_pipeline(target)
+        return ProjectBuilder(self._project, self).make_dataset(pipeline)
+
+    @property
+    def is_working_tree(self) -> bool:
+        return not self._rev
+
+    def source_data_dir(self, source) -> str:
+        if self.is_working_tree:
+            return self._project.source_data_dir(source)
+
+        obj_hash = self.build_targets[source].head.hash
+        return self._project.cache_path(obj_hash)
+
+
+class DiffStatus(Enum):
+    added = auto()
+    modified = auto()
+    removed = auto()
+    missing = auto()
+    foreign_modified = auto()
+
+Revision = NewType('Revision', str) # a commit hash or a named reference
+Reference = NewType('Reference', str) # a commit or an object hash
```

Comment:
`Revision` is everything that can be a Project Revision (which user directly works with - a commit hash, a branch, `HEAD`, a tag, a remote branch, BUT not the working tree), so it is any supported git single commit reference.

`Reference` is an object hash in the project cache. Either a resolved `Revision` (i.e., a Project Revision after rev parse), or a DVC object hash. Yes, it looks like it could be renamed to "Object id" without problems. `RefKind` just distinguishes between these two variants. `blob` - is a DVC object, `tree` - is the only possible kind of git objects in the project cache  (which is equivalent to a "revision tree" from the glossary). Maybe `tree` can be renamed to `commit`, but I'm not really sure it must be a commit and not its tree - because it could improve deduplication, but maybe introduce some other problems.

Reply:
Maybe you could change the commit on `ObjectId` to "Git commit or DVC object hash", then? That would improve clarity.

Reply:
Unfortunately, comments tend to be outdated.